runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -10393,227 +10393,637 @@ static TR::Register *inlineIsAssignableFrom(TR::Node *node, TR::CodeGenerator *c
+         //fall through
+      case TR::Int16:
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::vxor, node, high4Reg, high4Reg, high4Reg);
+         //fall through
+      case TR::Int32:
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::vxor, node, low4Reg, low4Reg, low4Reg);
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+      }
+   // ready all zeros and ones registers
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vxor, node, vconstant0Reg, vconstant0Reg, vconstant0Reg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vnor, node, vconstantNegReg, vconstant0Reg, vconstant0Reg);
+   // all four words in vunpackMaskReg a mask for the length of elementType
+   generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::vsldoi, node, vunpackMaskReg, vconstant0Reg, vconstantNegReg, TR::DataType::getSize(elementType));
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::vspltw, node, vunpackMaskReg, vunpackMaskReg, 3);
+   // vend = end & (~0xf) = end of aligned data
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, tempReg, 0xFFFFFFF0);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::AND, node, vendReg, endReg, tempReg);
+
+   // load the initial value
+   TR::Register *initialValueReg = low4Reg;
+   if (nonZeroInitial)
+      {
+      if (isLE) // in LE, load into the highest word of the highest vector
+         {
+         switch (elementType)
+            {
+            case TR::Int8:
+               initialValueReg = fourth4Reg;
+               break;
+            case TR::Int16:
+               initialValueReg = high4Reg;
+               break;
+            case TR::Int32:
+               //initialValueReg = low4Reg; by default
+               break;
+            default:
+               TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+            }
+         generateTrg1Src1Instruction(cg, TR::InstOpCode::mtvsrwz, node, initialValueReg, hashReg);
+         generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::vsldoi, node, initialValueReg, vconstant0Reg, initialValueReg, 8);
+         generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::vsldoi, node, initialValueReg, initialValueReg, vconstant0Reg, 12);
+         }
+      else // in BE, load into the lowest word
+         {
+         //initialValueReg = low4Reg; by default
+         generateTrg1Src1Instruction(cg, TR::InstOpCode::mtvsrwz, node, low4Reg, hashReg);
+         generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::vsldoi, node, low4Reg, vconstant0Reg, low4Reg, 8);
+         }
+      }
+
+   // if v is 16byte aligned, goto the VSX_Loop
+   loadConstant(cg, node, 0xF, tempReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::AND, node, tempReg, valueReg, tempReg);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, condReg, tempReg, 0x0);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::beq, node,
```

Comment:
Although its from the older code, but wondering if `and_r` (`and.`) instruction can be used to not need a separate `cmp` instruction, and if its worth it.